engine/buildkit/executor_spec.go

```diff
@@ -626,114 +616,136 @@ func (w *Worker) setupStdio(_ context.Context, state *execState) error {
-	if w.execMD != nil {
-		state.spec.Process.Env = append(state.spec.Process.Env, w.execMD.OTELEnvs...)
-	}
+const InstrumentationLibrary = "dagger.io/engine.buildkit"
 
-	if state.otelEndpoint == "" {
+func (w *Worker) setupOTel(ctx context.Context, state *execState) error {
+	if state.procInfo.Meta.NetMode != pb.NetMode_UNSET {
+		// align with setupNetwork; otherwise we hang waiting for a netNS worker
 		return nil
 	}
 
-	traceParent, ok := state.origEnvMap[OTELTraceParentEnv]
-	if ok {
-		otelCtx := propagation.TraceContext{}.Extract(ctx, propagation.MapCarrier{"traceparent": traceParent})
-		otelLogger := telemetry.Logger("dagger.io/executor")
-		stdout := &telemetry.OtelWriter{
-			Ctx:    otelCtx,
-			Logger: otelLogger,
-			Stream: 1,
-		}
-		stderr := &telemetry.OtelWriter{
-			Ctx:    otelCtx,
-			Logger: otelLogger,
-			Stream: 2,
-		}
-		state.procInfo.Stdout = nopCloser{io.MultiWriter(stdout, state.procInfo.Stdout)}
-		state.procInfo.Stderr = nopCloser{io.MultiWriter(stderr, state.procInfo.Stderr)}
+	logAttrs := []log.KeyValue{}
+	if w.execMD != nil {
+		state.spec.Process.Env = append(state.spec.Process.Env, w.execMD.OTelEnvs...)
+		logAttrs = append(logAttrs, log.String(telemetry.ClientIDAttr, w.execMD.ClientID))
 	}
+	logs := telemetry.Logs(ctx, InstrumentationLibrary, logAttrs...)
+
+	state.procInfo.Stdout = nopCloser{io.MultiWriter(logs.Stdout, state.procInfo.Stdout)}
+	state.procInfo.Stderr = nopCloser{io.MultiWriter(logs.Stderr, state.procInfo.Stderr)}
+	state.cleanups.add("close logs", logs.Close)
 
-	if strings.HasPrefix(state.otelEndpoint, "/") {
-		// Buildkit currently sets this to /dev/otel-grpc.sock which is not a valid
-		// endpoint URL despite being set in an OTEL_* env var.
-		state.otelEndpoint = "unix://" + state.otelEndpoint
+	listener, err := runInNetNS(ctx, state, func() (net.Listener, error) {
+		return net.Listen("tcp", "127.0.0.1:0")
+	})
+	if err != nil {
+		return fmt.Errorf("otel tcp proxy listen: %w", err)
 	}
 
-	if strings.HasPrefix(state.otelEndpoint, "unix://") && state.procInfo.Meta.NetMode == pb.NetMode_UNSET {
-		// if otel endpoint is unix and we are in CNI mode, setup tcp proxying of the unix endpoint for better client compatibility
-		if state.otelUnixSockSrcPath == "" {
-			return fmt.Errorf("no mount found for otel unix socket %s", state.otelUnixSockSrcPath)
-		}
+	proxyConnPool := pool.New()
+	listenerCtx, cancelListener := context.WithCancel(ctx)
+	listenerPool := pool.New().WithContext(listenerCtx).WithCancelOnError()
 
-		listener, err := runInNetNS(ctx, state, func() (net.Listener, error) {
-			return net.Listen("tcp", "127.0.0.1:0")
-		})
-		if err != nil {
-			return fmt.Errorf("otel tcp proxy listen: %w", err)
-		}
-		state.otelEndpoint = "http://" + listener.Addr().String()
+	otelProto := "http/protobuf"
+	otelEndpoint := "http://" + listener.Addr().String()
+	otelSrv := &http.Server{
+		Handler: w.telemetryPubSub,
```

Comment:
Neat, https://github.com/dagger/dagger/pull/7315 among other things makes the whole engine API just a single HTTP server, but needed to do some very light hijinks to still serve the gRPC otel endpoints: https://github.com/sipsma/dagger/blob/55a8e049ca22e7b8d76e86079e9b9e152d8b4465/cmd/engine/main.go#L424-L424

Would it make sense to now just install this pubsub http handler instead?

It's really not that big of a deal either way, checking the content type for grpc is ultimately fine, but would take that simplification if it makes sense.

This may require way too much context about the other PR to answer now in which case plz ignore; can figure it out later.

Reply:
Ha sweet, yeah that sounds plausible! I guess the only concern would be whether we want to also expose the full engine API, or just these OTLP endpoints, since this will be passed to everything regardless of whether they enabled nesting.